pkg/bpv7/extension_block_bcb_iop_aes_gcm.go

```diff
@@ -0,0 +1,483 @@
+			return fmt.Errorf("AES-128 variant %d and Keylength %d does not match", *aesVariantParameter, len(privateKey))
+		}
+	case 32:
+		if aesVariantParameter != nil && *aesVariantParameter != A256GCM {
+			return fmt.Errorf("AES-256 variant %d and Keylength %d does not match", *aesVariantParameter, len(privateKey))
+		}
+	default:
+		return fmt.Errorf("keylength %d is not supported", len(privateKey))
+	}
+	return nil
+}
+
+// EncryptTarget encrypts the target block using the BCB-IOP-AES-GCM security operation.
+func (bcb *BCBIOPAESGCM) EncryptTarget(b Bundle, bcbBlockNumber uint64, privateKey []byte) (err error) {
+
+	// Check if the target is a supported payload block
+	securityTargetBlock, err := b.GetExtensionBlockByBlockNumber(bcb.Asb.SecurityTargets[0])
+	if err != nil {
+		return err
+	}
+	if securityTargetBlock.Value.BlockTypeCode() != ExtBlockTypePayloadBlock {
+		return fmt.Errorf("unsupported security target block type code %d, %s", securityTargetBlock.Value.BlockTypeCode(), securityTargetBlock.Value.BlockTypeName())
+	}
+
+	// Remove CRC if present
+	if securityTargetBlock.CRCType != CRCNo {
+		securityTargetBlock.CRCType = CRCNo
+		securityTargetBlock.CRC = nil
+	}
+
+	// Extract the security target plaintext
+	plainText, err := bcb.extractPlainText(securityTargetBlock)
+	if err != nil {
+		return err
+	}
+
+	// Prepare the AAD
+	aad, err := bcb.prepareAAD(b, securityTargetBlock, bcbBlockNumber)
+	if err != nil {
+		return err
+	}
+
+	// Compute the cipherText and authenticationTag
+	cipherText, authenticationTag, err := bcb.computeAuthenticationTagAndCipherText(plainText, aad, privateKey)
+	if err != nil {
+		return err
+	}
+
+	// Set the cipherText as the payload
+	securityTargetBlock.Value = NewPayloadBlock(cipherText)
+
+	// Set the authenticationTag as security result
+	bcb.Asb.SecurityResults[0].results = append(bcb.Asb.SecurityResults[0].results, &IDValueTupleByteString{
+		id:    SecConResultIDBCBIOPAESGCMAuthenticationTag,
+		value: authenticationTag,
+	})
+
+	return nil
+
+}
+
+// DecryptTarget decrypts the payload of the security target block and verifies the authentication tag
+func (bcb *BCBIOPAESGCM) DecryptTarget(b Bundle, bcbBlockNumber uint64, privateKey []byte) (err error) {
+
+	// Check if the target is a supported payload block
+	securityTargetBlock, err := b.GetExtensionBlockByBlockNumber(bcb.Asb.SecurityTargets[0])
+	if err != nil {
+		return err
+	}
+
+	if securityTargetBlock.Value.BlockTypeCode() != ExtBlockTypePayloadBlock {
+		return fmt.Errorf("unsupported security target block type code %d, %s", securityTargetBlock.Value.BlockTypeCode(), securityTargetBlock.Value.BlockTypeName())
+	}
+
+	// Decrypt and Authenticate
+	plainText, err := bcb.decryptAndAuthenticate(b, securityTargetBlock, bcbBlockNumber, privateKey)
+	if err != nil {
+		return err
+	}
+
+	// Set the plainText as the payload
+	securityTargetBlock.Value = NewPayloadBlock(plainText)
+
+	// Set CRC
+	securityTargetBlock.CRCType = CRC32
```

Comment:
Why set this CRC type?

Reply:
The CRC Value is removed, according to the RFC, prior to encryption. Than the value has the be computed again after decryption, since AES-GCM also provides authentication, nothing is lost and also the CRC Value would be wrong since the payload is now encrypted, so it has a different CRC.
Does the CRC Type not have to be set after recomputing the CRC-Value?